File: swift/anagram/Anagram.swift
//
//  Anagram.swift
//  exercism
//
//  Created by Eric Bailey on 1/8/16.
//
//

import Foundation


/**
 Given a word and an array of possible anagrams, return the correct subarray.

 For example, given `"listen"` and an array of candidates,

 ```swift
 ["enlists", "google", "inlets", "banana"]
 ```

 ... return the singleton array,

 ```swift
 ["inlets"]
 ```

 This exercise was inspired by the [Extreme Startup game][1].

 [1]: https://github.com/rchatley/extreme_startup
 */
public struct Anagram {

  /// A `Word` for which to find anagrams.
  public let word: Word

  /**
   Create an `Anagram` with a given `word`.

   - parameter word: A `Word` for which to find anagrams.
   */
  public init(word: Word) {
    self.word = word
  }

  /**
   Given an array of `candidates`, return the possibly empty array
   of its elements that are anagrams of `word`.

   - parameter candidates: An array of potential anagrams of `word`.
   - returns: The subarray of `candidates` which are anagrams of `word`.
   */
  public func match(candidates: [Word]) -> [Word] {
    return candidates.filter { $0.lowercaseString.isAnagram(word.lowercaseString) }
  }

}

public extension Word {

  /**
   Given a lowercased word, return `true` iff `self`
   is not equal to and has the same letters as `word`.

   - parameter word: A `Word` of which `self` might be an anagram.
   - returns: `true` iff `self` is an anagram of `word`.

   - requires: `self` and `word` must be lowercased.
   - seealso: `hasSameLetters(_:)`
   */
  public func isAnagram(word: Word) -> Bool {
    return self != word && self.hasSameLetters(word)
  }

  /**
   Given a lowercased word, return `true` iff `self` has the same letters as `word`.

   - parameter word: A word with which to compare `self`.
   - returns: `true` iff `self` has the same letters as `word`.

   - requires: `self` and `word` must be lowercased.
   */
  public func hasSameLetters(word: Word) -> Bool {
    return characters.sort() == word.characters.sort()
  }

}

File: go/anagram/anagram.go
// Package anagram implements the Anagram exercise from Exercism.
package anagram

import (
	"sort"
	"strings"
)

// Detect returns the subset of candidates that are anagrams of a given target.
func Detect(target string, candidates []string) (anagrams []string) {
	targetLowered := strings.ToLower(target)
	targetSortedRunes := sortedRunes(targetLowered)

	for _, candidate := range candidates {
		if isAnagram(targetLowered, targetSortedRunes, candidate) {
			anagrams = append(anagrams, candidate)
		}
	}
	return
}

func isAnagram(targetLowered string, targetSortedRunes []rune, candidate string) bool {
	candidateLowered := strings.ToLower(candidate)
	if targetLowered == candidateLowered {
		return false
	}

	candidateSortedRunes := sortedRunes(candidateLowered)
	if len(targetSortedRunes) != len(candidateSortedRunes) {
		return false
	}

	for i, r := range targetSortedRunes {
		if candidateSortedRunes[i] != r {
			return false
		}
	}

	return true
}

func sortedRunes(word string) []rune {
	sorted := []rune(word)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i] < sorted[j]
	})
	return sorted
}
